File: granular-engine/src/engine.rs
use num_rational::Rational64;

use crate::{
    buffer::Buffer,
    env::Env,
    time,
    voice_manager::{VoiceManager, VoiceMode},
};

pub struct Engine<const S: usize> {
    sample_rate: usize,
    params: EngineParams,
    last_tick: u64,
    last_note_event: i32,
    voices: VoiceManager<6, S>,
}

impl<const S: usize> Engine<S> {
    pub fn new(cfg: EngineConfig, params: EngineParams) -> Self {
        let EngineParams {
            bpm,
            attack_ms,
            decay_ms,
            sustain,
            release_ms,
        } = params;
        let voices = VoiceManager::new(
            VoiceMode::Poly,
            time::ms_to_samples(cfg.sample_rate, attack_ms),
            time::ms_to_samples(cfg.sample_rate, decay_ms),
            sustain,
            time::ms_to_samples(cfg.sample_rate, release_ms),
            time::bpm_to_freq(bpm),
            cfg.sample_rate,
        );
        Self {
            sample_rate: cfg.sample_rate,
            params,
            last_tick: 0,
            last_note_event: 0,
            voices,
        }
    }

    pub fn reset_after_update_sample(&mut self) {
        self.voices.reset_grains();
    }

    pub fn set_bpm(&mut self, bpm: u32) {
        if bpm == self.params.bpm {
            return;
        }
        self.voices.set_lead_clock_freqs(time::bpm_to_freq(bpm));
        self.params.bpm = bpm;
    }

    pub fn set_adsr(&mut self, attack_ms: u32, decay_ms: u32, sustain: f32, release_ms: u32) {
        self.params.attack_ms = attack_ms;
        self.params.decay_ms = decay_ms;
        self.params.sustain = sustain;
        self.params.release_ms = release_ms;
        self.voices.set_adsr(
            time::ms_to_samples(self.sample_rate, attack_ms),
            time::ms_to_samples(self.sample_rate, decay_ms),
            sustain,
            time::ms_to_samples(self.sample_rate, release_ms),
        );
    }

    pub fn set_stream_enabled(&mut self, stream_id: usize, enabled: bool) {
        self.voices.set_stream_enabled(stream_id, enabled);
    }

    pub fn set_stream_subdivision(&mut self, stream_id: usize, subdivision: u32) {
        self.voices
            .subdivide_stream_clock(stream_id, Rational64::ONE * subdivision as i64);
    }

    pub fn set_stream_grain_start(&mut self, stream_id: usize, grain_start: f32) {
        self.voices.set_stream_grain_start(stream_id, grain_start);
    }

    pub fn set_stream_grain_size_ms(&mut self, stream_id: usize, grain_size_ms: usize) {
        self.voices
            .set_stream_grain_size_ms(stream_id, grain_size_ms);
    }

    pub fn set_stream_grain_probability(&mut self, stream_id: usize, probability: f32) {
        self.voices
            .set_stream_grain_probability(stream_id, probability);
    }

    pub fn set_stream_gain(&mut self, stream_id: usize, gain: f32) {
        self.voices.set_stream_gain(stream_id, gain);
    }

    pub fn set_stream_tune(&mut self, stream_id: usize, tune: i32) {
        self.voices.set_stream_tune(stream_id, tune);
    }

    pub fn set_stream_pan(&mut self, stream_id: usize, pan: f32) {
        self.voices.set_stream_pan(stream_id, pan);
    }

    pub fn set_stream_env(&mut self, stream_id: usize, env: Env) {
        self.voices.set_stream_env(stream_id, env);
    }

    /// Positive-valued `note_event`s indicate a note-on event for the MIDI note
    /// with the value `note_event - 1`. Negative-valued events indicate
    /// note-off events for abs(note_event) - 1.
    pub fn apply_note_event(&mut self, note_event: i32) {
        if note_event == 0 || note_event == self.last_note_event {
            return;
        }

        self.last_note_event = note_event;

        let note = note_event.abs() as u32 - 1;
        if note_event > 0 {
            self.voices.note_on(note, self.last_tick);
        } else {
            self.voices.note_off(note);
        }
    }

    pub fn process(
        &mut self,
        sample_buf: &Buffer,
        output_buf: &mut Buffer,
        playheads_buf: &mut Buffer,
    ) {
        for i in 0..output_buf.len {
            let mut frame = [0., 0.];

            // TODO - polyphonic playhead positions
            let mut playhead_positions = [-1.; S];
            self.voices.spawn_new_grains(&sample_buf, self.sample_rate);
            self.voices
                .render_frame(sample_buf, &mut frame, &mut playhead_positions);
            self.voices.tick();

            output_buf[0][i] = frame[0];
            output_buf[1][i] = frame[1];
            for (s, pos) in playhead_positions.iter().enumerate() {
                playheads_buf[s][i] = *pos;
            }

            self.last_tick += 1;
        }
    }
}

pub struct EngineConfig {
    pub sample_rate: usize,
    pub output_buf_len: usize,
    pub output_buf_capacity: usize,
}

pub struct EngineParams {
    pub bpm: u32,
    pub attack_ms: u32,
    pub decay_ms: u32,
    pub sustain: f32,
    pub release_ms: u32,
}

impl Default for EngineParams {
    fn default() -> Self {
        Self {
            bpm: 60,
            attack_ms: 100,
            decay_ms: 50,
            sustain: 0.5,
            release_ms: 250,
        }
    }
}

File: granular-app/src/synth/granular/engine/Engine.ts
import * as Buf from "../../../lib/Buffer";
import { DefaultLogger } from "../../../lib/DefaultLogger";
import { type Logger } from "../../../lib/Logger";
import { Config } from "../Config";
import {
  StreamParamKey,
  unpackStreamParam,
  type GranularParams,
} from "../param";
import { type Pointer } from "./Exports";
import { Instance } from "./Instance";

/**
 * Wrapper class providing safe access to an instance of the WASM granular
 * engine
 */
export class Engine {
  private readonly instance: Instance;
  private readonly engine: Pointer;
  private readonly outputBuf: Pointer;
  private outputBufView: Buf.Buffer = Buf.create(2);
  private readonly playheadsBuf: Pointer;
  private playheadsBufView: Buf.Buffer = Buf.create(Config.NumStreams);
  private sampleBuf?: Pointer;
  private readonly log: Logger;

  private audioBufLen: number;
  private audioBufCapacity: number;

  constructor(
    module: WebAssembly.Module,
    options: {
      sampleRate: number;
      outputBufCapacity: number;
      outputBufLen: number;
    },
  ) {
    this.log = new DefaultLogger(Engine.name);

    this.instance = new Instance(module, {
      Math: {
        random: Math.random,
      },
    });

    this.audioBufCapacity = options.outputBufCapacity;
    this.audioBufLen = options.outputBufLen;

    this.engine = this.instance.exports.new_engine(
      sampleRate,
      this.audioBufLen,
      this.audioBufCapacity,
    );
    this.outputBuf = this.instance.exports.new_buffer(
      2,
      this.audioBufLen,
      this.audioBufCapacity,
    );
    this.playheadsBuf = this.instance.exports.new_buffer(
      Config.NumStreams,
      this.audioBufLen,
      this.audioBufCapacity,
    );

    this.createBufferViews();
  }

  setParams(params: GranularParams) {
    const {
      bpm,
      note_event,
      attack,
      decay,
      sustain,
      release,
      ...streamParams
    } = params;
    this.instance.exports.set_bpm(this.engine, bpm[0]);
    this.instance.exports.apply_note_event(this.engine, note_event[0]);
    this.instance.exports.set_adsr(
      this.engine,
      attack[0],
      decay[0],
      sustain[0],
      release[0],
    );

    for (const [streamParam, [val]] of Object.entries(streamParams)) {
      const result = unpackStreamParam(streamParam as StreamParamKey);
      if (!result) {
        this.log.warn("unable to parse StreamParam", { streamParam });
        continue;
      }
      const [streamId, param] = result;
      switch (param) {
        case "enabled":
          this.instance.exports.set_stream_enabled(this.engine, streamId, val);
          break;
        case "subdivision":
          this.instance.exports.set_stream_subdivision(
            this.engine,
            streamId,
            val,
          );
          break;
        case "grainStart":
          this.instance.exports.set_stream_grain_start(
            this.engine,
            streamId,
            val,
          );
          break;
        case "grainSizeMs":
          this.instance.exports.set_stream_grain_size_ms(
            this.engine,
            streamId,
            val,
          );
          break;
        case "grainProbability":
          this.instance.exports.set_stream_grain_probability(
            this.engine,
            streamId,
            val,
          );
          break;
        case "gain":
          this.instance.exports.set_stream_gain(this.engine, streamId, val);
          break;
        case "tune":
          this.instance.exports.set_stream_tune(this.engine, streamId, val);
          break;
        case "pan":
          this.instance.exports.set_stream_pan(this.engine, streamId, val);
          break;
        case "env":
          this.instance.exports.set_stream_env(this.engine, streamId, val);
          break;
        default:
          this.log.warn(`unexpected param type ${param}`);
          break;
      }
    }
  }

  process(): [Buf.Buffer, Buf.Buffer] {
    if (typeof this.sampleBuf !== "undefined") {
      this.instance.exports.process(
        this.engine,
        this.sampleBuf,
        this.outputBuf,
        this.playheadsBuf,
      );
    }

    return [this.outputBufView, this.playheadsBufView];
  }

  updateSample(sample: Buf.Buffer) {
    this.log.info("allocating sample buffer");
    const bufLen = Buf.length(sample);
    if (typeof this.sampleBuf === "undefined") {
      this.sampleBuf = this.instance.exports.new_buffer(2, bufLen, bufLen);
    } else {
      this.instance.exports.resize_buffer(this.sampleBuf, bufLen, bufLen);
    }

    // If our allocation causes the WASM memory to grow, we will have to re-draw
    // our views over its memory buffer. The WASM memory model guarantees that
    // our pointers are still accurate relative to the buffer, but the buffer
    // may itself have moved.
    this.createBufferViews();
    const destBufView = [
      this.channelView(
        this.instance.exports.buffer_channel(this.sampleBuf, 0),
        bufLen,
      ),
      this.channelView(
        this.instance.exports.buffer_channel(this.sampleBuf, 1),
        bufLen,
      ),
    ];
    Buf.copyStereo(sample, destBufView);
    this.instance.exports.reset_after_update_sample(this.engine);
  }

  // Based on the way the WASM memory model works, we cannot simply pass a
  // pointer to the WebAudio-provided `outputs` buffers from the
  // `AudioWorkletProcessor.process()` callback, which would have been
  // convenient for us, since we wouldn't have to allocate anything at
  // runtime, at least from within code we control.
  //
  // Instead, we're stuck allocating and potentially growing buffers within
  // the WASM instance's memory, i.e. if we get called back with a `x >
  // 128`-length buffer to fill (which is unlikely to happen, but Web Audio
  // claims they will build this out eventually).
  //
  // We have a couple mitigations. First, WASM instances are created with a
  // pre-allocated block of growable linear memory, meaning that in the event
  // of a small-enough buffer allocation, the WASM runtime won't be
  // `malloc`ing anything, but rather just moving pointers around. When the
  // WASM runtime needs to actually grow its linear memory, allocations are
  // done in 64 KiB pages. I've never seen an audio buffer size larger than
  // 8192 f32s, so it's unlikely we'll ever require a real heap allocation
  // more than once.
  //
  // When we do resize a buffer, we allocate a capacity that's roughly 2x the
  // buffer size we see, KiB-aligned, in the event that we resize again.
  checkResizeProcessingBuffers(bufLen: number) {
    if (bufLen <= this.audioBufCapacity) {
      return;
    }

    this.audioBufLen = bufLen;
    this.audioBufCapacity = Math.ceil(this.audioBufLen / 1024) * 1024;
    this.instance.exports.resize_buffer(
      this.outputBuf,
      this.audioBufCapacity,
      this.audioBufLen,
    );
    this.instance.exports.resize_buffer(
      this.playheadsBuf,
      this.audioBufCapacity,
      this.audioBufLen,
    );

    this.createBufferViews();
  }

  private createBufferViews() {
    this.log.info("(re)creating buffer views");

    this.outputBufView[0] = this.channelView(
      this.instance.exports.buffer_channel(this.outputBuf, 0),
      this.audioBufLen,
    );
    this.outputBufView[1] = this.channelView(
      this.instance.exports.buffer_channel(this.outputBuf, 1),
      this.audioBufLen,
    );

    for (let i = 0; i < Config.NumStreams; i++) {
      this.playheadsBufView[i] = this.channelView(
        this.instance.exports.buffer_channel(this.playheadsBuf, i),
        this.audioBufLen,
      );
    }
  }

  private channelView(ptr: Pointer, len: number): Float32Array {
    return new Float32Array(this.instance.exports.memory.buffer, ptr, len);
  }
}
